Moved coverage counter injection from BasicBlock to Statement.

richkadel opened the pull request:
As discussed on Zulip: https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Implement.20LLVM-compatible.20source-based.20cod.20compiler-team.23278

Commit:
Moved coverage counter injection from BasicBlock to Statement.

rust-highfive commented:
r? @ecstatic-morse

(rust_highfive has picked a reviewer for you, use r? to override)

richkadel commented:
r? @wesleywiser 
FYI @tmandry

richkadel commented:
This code builds for me without errors, even after I did another `./x.py clean`. So I don't see why it's failing in CI with errors related to the `derive...RustcEncodable, RustcDecodable...`.

I made a couple of what seem to be insignificant changes just now, and pushed them after successful testing locally. We'll see if something changes.

Note the way that I'm deriving these traits appears to be identical to how they are derived for other structs and enums, even in the same crate `librustc_middle/mir/mod.rs`, but the only failures are related to the new structs and enums that I just added.

Maybe I need a bootstrap config? I'm not sure.

Here's a snippet from the errors in the CI log, for reference:

```
    Checking rustc_builtin_macros v0.0.0 (/checkout/src/librustc_builtin_macros)
error[E0107]: wrong number of type arguments: expected 1, found 0
    --> src/librustc_middle/mir/mod.rs:1510:51
     |
1510 |   #[derive(Clone, Debug, PartialEq, RustcEncodable, RustcDecodable, HashStable, TypeFoldable)]
     |                                                     ^^^^^^^^^^^^^^
     |                                                     |
     |                                                     expected 1 type argument
     |                                                     in this macro invocation
     | 
    ::: /checkout/library/core/src/macros/mod.rs:1432:5
     |
1432 | /     pub macro RustcDecodable($item:item) {
1433 | |         /* compiler built-in */
1434 | |     }
     | |_____- in this expansion of `#[derive(RustcDecodable)]`

error[E0107]: wrong number of type arguments: expected 1, found 0
    --> src/librustc_middle/mir/mod.rs:1510:35
     |
1510 |   #[derive(Clone, Debug, PartialEq, RustcEncodable, RustcDecodable, HashStable, TypeFoldable)]
     |                                     ^^^^^^^^^^^^^^
     |                                     |
     |                                     expected 1 type argument
```

richkadel commented:
I just tested this on my local Linux workstation, after completely deleting my `build` directory first, and I can't reproduce this error.

@wesleywiser @tmandry - Are you familiar with this kind of issue?

The only thing different is I synched up my fork from rust-lang/rust yesterday around noon.

I can try synching again, but it seems too high of a coincidence that someone changed something in RustcEncoder/Decoder in the last 24 hours. Plus the error is only affecting my new structs and enums, and none of the others in the same package.

So I assume there's a different explanation.

Mark-Simulacrum commented:
https://github.com/rust-lang/rust/pull/73851 merged 16 hours ago and is a pretty big rework of the serialization infra, I would rebase atop that.

richkadel commented:
Ah. Thanks Mark. I'll do that.

richkadel commented:
OK, yes, I can repro this locally now.

richkadel commented on src/test/mir-opt/instrument_coverage.bar.InstrumentCoverage.diff:
```diff
@@ -72,6 +14,10 @@
           return;                          // scope 0 at /the/src/instrument_coverage.rs:21:2: 21:2
++     }
++ 
++     bb1 (cleanup): {
```
I think this might have been added by the patch implementation. Since I'm only inserting a statement, maybe I don't need to use patch at all.

richkadel replied:
Fixed

RalfJung commented on src/librustc_mir/interpret/step.rs:
```diff
@@ -117,6 +117,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
             // Statements we do not track.
             AscribeUserType(..) => {}
+            Coverage(..) => {}
```
This makes that statement a NOP for Miri/CTFE. So it is always correct to just do nothing here, and there is no UB that can possibly be caused by these statements?

richkadel replied:
Thanks for the callout. Good question.

The referenced line of code is a great place for me to re-insert the FIXME comment I had removed from rustc:mir::interpret::intrinsics. I just [pushed a change adding some comments here, for `step`](https://github.com/rust-lang/rust/pull/75563/files#diff-fabaa9592cbf2a32d5acb6e8cc047908).

Correct me if I'm wrong, but I believe it is reasonable to treat the Coverage statement as a NOP, if evaluated in a const expression, at this time. The statement is injected by a MIR pass, in a way that should not have any side effects on the runtime behavior, so ignoring it here effectively "un-injects" it, as if it never existed (right?).

If the Coverage statement is a "Counter" kind, it would normally result in LLVM-generated runtime code of only three statements: load a global int, increment it, and store it. 

If and/or what we might want to do in the future is the subject of the FIXME issue #73156.

RalfJung replied:
> Correct me if I'm wrong, but I believe it is reasonable to treat the Coverage statement as a NOP, if evaluated in a const expression, at this time. The statement is injected by a MIR pass, in a way that should not have any side effects on the runtime behavior, so ignoring it here effectively "un-injects" it, as if it never existed (right?).

Yes, that sounds right.

Note that your comment only seems to take into account CTFE. This code here is also used by [the Miri UB checker](https://github.com/rust-lang/miri/). So it has to adequately model all UB. But since coverage does not exist on the source level, and since presumably it is not something the program can interact with directly, I assume there also is no UB to check for here.

The comment should be adjusted taking into account that this is not just for const expressions. This is for all code that is evaluated with the Miri engine.

richkadel replied:
Got it. I'll make that change. Thanks.

richkadel replied:
done

RalfJung replied:
Looking good, thanks. :)

richkadel commented:
Thanks for the callout. Good question.

The referenced line of code is a great place for me to re-insert the FIXME comment I had removed from rustc:mir::interpret::intrinsics. I just [pushed a change adding some comments here, for `step`](https://github.com/rust-lang/rust/pull/75563/files#diff-fabaa9592cbf2a32d5acb6e8cc047908).

Correct me if I'm wrong, but I believe it is reasonable to treat the Coverage statement as a NOP, if evaluated in a const expression, at this time. The statement is injected by a MIR pass, in a way that should not have any side effects on the runtime behavior, so ignoring it here effectively "un-injects" it, as if it never existed (right?).

If the Coverage statement is a "Counter" kind, it would normally result in LLVM-generated runtime code of only three statements: load a global int, increment it, and store it. 

If and/or what we might want to do in the future is the subject of the FIXME issue #73156.

richkadel commented:
@Mark-Simulacrum - I'd like to "@bors try" this PR with tests enabled, for Linux, Mac, and Windows-Gnu. (MSVC is not currently supported).

Can you provide current instructions to set that up?

Thanks!

Mark-Simulacrum commented:
You should enable at most 2 builders on try at the same time (and wait to kick off a build until the previous one finishes), with at most 1 apple builder in any try run.

The steps are roughly:
* For linux and windows: edit the try builder list [here](https://github.com/rust-lang/rust/blob/e8df0b893235aa06c6362d1a8646296310f79f81/src/ci/github-actions/ci.yml#L296-L297) to include whichever builders you want, pulling from the rest of that file, e.g., in this case you'll want to look for `name: x86_64-apple` for the apple tests, `name: x86_64-mingw-1` or `name: x86_64-mingw-2` (2 runs UI and compile-fail tests, 1 runs everything else), and `name: x86_64-gnu` for main linux testing.

./x.py run src/tools/expand-yaml-anchors

bors commented:
:umbrella: The latest upstream changes (presumably #75187) made this pull request unmergeable. Please resolve the merge conflicts.

richkadel commented:
@Mark-Simulacrum - Thanks! I'm starting with Linux and MacOS (1 builder each).

See below. If I've done this wrong, let me know.

When you say _"...and wait to kick off a build until the previous one finishes", do you mean, if there is another _try_ job currently in progress, manually watch the queue, wait for it to finish, and then submit `@bors try` request, hopefully getting it in before someone else?

Or is there a way I can automate this?

(FYI, I'm not sure I have the github permissions to start my own `@bors try`. I may need @wesleywiser or @tmandry to do it.)

```yml
    strategy:
      matrix:
        include:
          - name: x86_64-gnu
            <<: *job-linux-xl
          - name: x86_64-apple
            env:
              SCRIPT: ./x.py --stage 2 test
              RUST_CONFIGURE_ARGS: --build=x86_64-apple-darwin --enable-sanitizers --enable-profiler --set rust.jemalloc
              RUSTC_RETRY_LINKER_ON_SEGFAULT: 1
              MACOSX_DEPLOYMENT_TARGET: 10.8
              MACOSX_STD_DEPLOYMENT_TARGET: 10.7
              NO_LLVM_ASSERTIONS: 1
              NO_DEBUG_ASSERTIONS: 1
            <<: *job-macos-xl
```

richkadel commented:
(FYI, if the snippet looks right, the PR has been updated, and is ready for the `@bors try`

tmandry commented:
@bors try

bors commented:
:hourglass: Trying commit 6381f8a4dcf3caf223b4289d465e932ae4f46de5 with merge 67ed07aa158e60bc073ab0ad5616071768ee2a5b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6381f8a4dcf3caf223b4289d465e932ae4f46de5","merge_sha":"67ed07aa158e60bc073ab0ad5616071768ee2a5b"} -->

Mark-Simulacrum commented:
Snippet looks right. Mostly it should be fine to start these on *this* PR not in parallel (i.e., don't request another try build on new push until this one finishes).

bors commented:
:umbrella: The latest upstream changes (presumably #75145) made this pull request unmergeable. Please resolve the merge conflicts.

richkadel commented:
@Mark-Simulacrum - The `try` has been running for quite a while.

I'm not able to find the running logs for my `try` jobs. Is there a way for me to see what's going on?

Thanks

Mark-Simulacrum commented:
I am uncertain what went wrong - it usually only takes 2 hours. It is running, though, and seems to be not idle at least: https://github.com/rust-lang-ci/rust/runs/995511136

Presumably that'll finish soon-ish.

richkadel commented:
OK, thanks.

Thanks for the link.

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

richkadel commented:
ha! test just timed out... 🤒

richkadel commented:
@tmandry - I have to fix a new merge conflict.

It looks like Linux tests succeeded, so I'll remove them and add Windows-gnu, then we can re-start try to hopefully pass off MacOS and Windows Gnu.

richkadel commented:
@tmandry - I think it's ready. Here's the change (and I ran `./x.py run src/tools/expand-yaml-anchors` to update the `.github/` file as well):

https://github.com/rust-lang/rust/pull/75563/files#diff-873868365228a2c21bc5301d6bb9920b

richkadel commented:
Actually, it looks like MacOS succeeded after all, despite the timeout.

If it's not too late, @tmandry, I'm going to change this to only try Windows.

wesleywiser commented:
@bors try

tmandry commented:
@bors try

bors commented:
:hourglass: Trying commit c4dc2b44024d03f149517a2bcb6845428cb36fe1 with merge 7a29c1aac47f0e285c9f05d75491506ebfd0b93c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c4dc2b44024d03f149517a2bcb6845428cb36fe1","merge_sha":"7a29c1aac47f0e285c9f05d75491506ebfd0b93c"} -->

tmandry commented:
oops, you win @wesleywiser

tmandry commented:
@richkadel just saw your comment as well, not sure if I can cancel a try run (maybe try- ?), but it's okay if we let it run I think.

wesleywiser commented on src/librustc_middle/mir/visit.rs:
```diff
@@ -134,6 +134,13 @@ macro_rules! make_mir_visitor {
 
+            fn visit_coverage(&mut self,
+                              coverage: & $($mutability)? Coverage,
+                              _context: PlaceContext,
```
I don't think we want/need `PlaceContext` here. That's used when there is a `Place` being visited which we don't have.

richkadel replied:
done/removed

richkadel commented:
The tries will probably fail quickly. There was an automated merge error with the last rebase. 

I fixed the error, and also removed apple.

Should be good to try now. 

Thanks!

wesleywiser commented on src/test/mir-opt/instrument_coverage.bar.InstrumentCoverage.diff:
```diff
@@ -3,74 +3,16 @@
                                            // mir::Constant
                                            // + span: /the/src/instrument_coverage.rs:20:5: 20:9
                                            // + literal: Const { ty: bool, val: Value(Scalar(0x01)) }
++         Coverage { kind: Counter { function_source_hash: 10208505205182607101, id: CounterValueReference(0) }, region: CodeRegion { file_name: "/the/src/instrument_coverage.rs", start_line: 19, start_col: 18, end_line: 21, end_col: 2 } }; // scope 0 at /the/src/instrument_coverage.rs:21:2: 21:2
```
It would be nice to make this a little prettier. Using the standard line:column notation like `19:18 - 21:2` would help a bit.

richkadel replied:
done

wesleywiser reviewed (commented):
Overall, I'm quite pleased with the simplifications this has allowed!

rust-log-analyzer commented:
The job `x86_64-mingw-2` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/996325181) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/7a29c1aac47f0e285c9f05d75491506ebfd0b93c/checks/996325181/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/7a29c1aac47f0e285c9f05d75491506ebfd0b93c/checks/996325181/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
   Compiling rustc_passes v0.0.0 (D:\a\rust\rust\src\librustc_passes)
   Compiling rustc_mir v0.0.0 (D:\a\rust\rust\src\librustc_mir)
   Compiling rustc_ty v0.0.0 (D:\a\rust\rust\src\librustc_ty)
   Compiling rustc_codegen_llvm v0.0.0 (D:\a\rust\rust\src\librustc_codegen_llvm)
error: unused import: `tracing::debug`
  --> src\librustc_codegen_llvm\intrinsic.rs:26:5
26 | use tracing::debug;
   |     ^^^^^^^^^^^^^^
   |
   |
   = note: `-D unused-imports` implied by `-D warnings`
error: aborting due to previous error

error: could not compile `rustc_codegen_llvm`.


To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "D:\\a\\rust\\rust\\build\\x86_64-pc-windows-gnu\\stage0\\bin\\cargo.exe" "build" "--target" "x86_64-pc-windows-gnu" "-Zbinary-dep-depinfo" "-j" "8" "--release" "--locked" "--color" "always" "--features" " llvm" "--manifest-path" "D:\\a\\rust\\rust\\src/rustc/Cargo.toml" "--message-format" "json-render-diagnostics"
expected success, got: exit code: 101
failed to run: D:\a\rust\rust\build\bootstrap\debug\bootstrap test --stage 2 src/test/ui src/test/compile-fail
Build completed unsuccessfully in 0:16:49
make: *** [Makefile:84: ci-mingw-subset-2] Error 1
  local time: Tue Aug 18 02:05:34 CUT 2020
  network time: Tue, 18 Aug 2020 02:05:34 GMT
== end clock drift check ==
== end clock drift check ==
##[error]Process completed with exit code 2.
Terminate orphan process: pid (7500) (python)
Terminate orphan process: pid (7268) (sccache)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

richkadel commented:
Good. It failed pretty quickly.

Can someone restart the `@bors try` with the fixes I last pushed?

Thanks!

richkadel commented:
@wesleywiser - Thanks for the review!

I wanted to make sure you saw [my Zulip comment today](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/Implement.20LLVM-compatible.20source-based.20cod.20compiler-team.23278/near/207208336).

Do you want me to leave this as is, with `StatementKind::Coverage(Box<Coverage>)`, or do you like the idea of `StatementKind::Injected(Box<InjectedKind>)`? (or something else)

tmandry commented:
@bors try

bors commented:
:hourglass: Trying commit 36a96633aa7982177668f7bc607364cf3c666f22 with merge 56bc179652e0615fe28be47135c9b5b7c6589ab5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"36a96633aa7982177668f7bc607364cf3c666f22","merge_sha":"56bc179652e0615fe28be47135c9b5b7c6589ab5"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/996666595), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36058)
Build commit: 56bc179652e0615fe28be47135c9b5b7c6589ab5 (`56bc179652e0615fe28be47135c9b5b7c6589ab5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36058","checks-actions":"https://github.com/rust-lang-ci/rust/runs/996666595"},"merge_sha":"56bc179652e0615fe28be47135c9b5b7c6589ab5"} -->

richkadel commented:
@wesleywiser I made changes per your feedback, and also removed the changes to ci.yml files that expanded the bors try tests.

bors commented:
:umbrella: The latest upstream changes (presumably #75566) made this pull request unmergeable. Please resolve the merge conflicts.

wesleywiser commented:
@richkadel 

> Do you want me to leave this as is, with StatementKind::Coverage(Box<Coverage>), or do you like the idea of StatementKind::Injected(Box<InjectedKind>)? (or something else)

I think for now, I'd prefer to leave it as `StatementKind::Coverage`. If we find a strong need to generalize this later, we can do that refactoring then.

richkadel commented:
I just resolved the new conflicts in the test .diff files.

@wesleywiser - If you're ready, can we submit to bors with norollup? There have been a lot of conflicts over the past 24 hours.

Thanks

richkadel commented:
Hmm, I must have missed the formatting issue last night. Fixed.

wesleywiser commented:
cc @rust-lang/compiler 

This PR adds a new MIR `StatementKind` which is used by the [LLVM code coverage feature](https://github.com/rust-lang/compiler-team/issues/278). Prior to this change, the approach used a few intrinsics to set up the code coverage instrumentation correctly but this required relatively complex modifications to generated MIR during the code coverage pass. 

The new approach here is much simpler overall and most of the rest of rustc can treat this new statement as a nop.

richkadel commented:
(Minor change. I just removed an unused parameter, is_cleanup, not needed now that I'm using "Statement".)

wesleywiser commented:
@bors r+

bors commented:
:pushpin: Commit d129ac2655451ca894bceeb58a4d5c26a86c4caa has been approved by `wesleywiser`

<!-- @bors r=wesleywiser d129ac2655451ca894bceeb58a4d5c26a86c4caa -->
<!-- homu: {"type":"Approved","sha":"d129ac2655451ca894bceeb58a4d5c26a86c4caa","approver":"wesleywiser"} -->

bors commented:
:hourglass: Testing commit d129ac2655451ca894bceeb58a4d5c26a86c4caa with merge 5f6fcad6449aabad3c29c65c1738f8eab89c03ce...
<!-- homu: {"type":"BuildStarted","head_sha":"d129ac2655451ca894bceeb58a4d5c26a86c4caa","merge_sha":"5f6fcad6449aabad3c29c65c1738f8eab89c03ce"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1005462808), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36104)
Approved by: wesleywiser
Pushing 5f6fcad6449aabad3c29c65c1738f8eab89c03ce to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36104","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1005462808"},"merge_sha":"5f6fcad6449aabad3c29c65c1738f8eab89c03ce"} -->

richkadel commented:
🎉 Thanks @wesleywiser and @tmandry ! This will make everything else much better and easier.

I'm making good progress on injecting counters with good coverage spans, for branches (starting with specific cases). I'll share these soon, along with tests.

tmiasko commented on src/librustc_middle/mir/visit.rs:
```diff
@@ -1133,6 +1150,8 @@ pub enum NonUseContext {
     StorageDead,
     /// User type annotation assertions for NLL.
     AscribeUserTy,
+    /// Coverage code region and counter metadata.
+    Coverage,
```
@richkadel is `PlaceContext::NonUse(NonUseContext::Coverage)` reserved for a future use? It's never constructed currently.

wesleywiser replied:
@tmiasko I think this got left in accidentally from a prior version of these changes where it was used. I believe it can be removed.
